30-indoor-air-quality-and-gas-leak/firmware/sparrow-application/aqi/aqi.cpp

```diff
@@ -0,0 +1,496 @@
+            }
+            else if (shouldSendMonitoringEvent()) {
+                addNote(sensorReading, false, ALERT_NONE);
+                schedSetCompletionState(appID, STATE_AQI_MONITORING, STATE_AQI_MONITORING);
+            }
+        }
+        break;
+
+    case STATE_AQI_ALERTING:
+        if (takeReading(sensorReading)) {
+            if (sensorReading.isAlert()) {
+                if (shouldSendMonitoringEvent()) {
+                    addNote(sensorReading, false, ALERT_ONGOING);
+                    schedSetCompletionState(appID, STATE_AQI_ALERTING, STATE_AQI_ALERTING);
+                }
+            }
+            else {
+                addNote(sensorReading, true, ALERT_STANDDOWN);
+                schedSetCompletionState(appID, STATE_AQI_MONITORING, STATE_AQI_ALERTING);
+            }
+        }
+        break;
+    }
+}
+
+// Register the notefile template for our data
+static bool registerNotefileTemplate()
+{
+
+    // Create the request
+    J *req = NoteNewRequest("note.template");
+    if (req == NULL) {
+        return false;
+    }
+
+    // Create the body
+    J *body = JCreateObject();
+    if (body == NULL) {
+        JDelete(req);
+        return false;
+    }
+
+    // Add an ID to the request, which will be echo'ed
+    // back in the response by the notecard itself.  This
+    // helps us to identify the asynchronous response
+    // without needing to have an additional state.
+    JAddNumberToObject(req, "id", REQUESTID_TEMPLATE);
+
+    // Fill-in request parameters.  Note that in order to minimize
+    // the size of the over-the-air JSON we're using a special format
+    // for the "file" parameter implemented by the gateway, in which
+    // a "file" parameter beginning with * will have that character
+    // substituted with the textified application address.
+    JAddStringToObject(req, "file", APPLICATION_NOTEFILE);
+
+    // Fill-in the body template
+    JAddStringToObject(body, DATA_FIELD_APP, TSTRINGV);
+    JAddNumberToObject(body, DATA_FIELD_ALERT, TUINT8);
+    JAddNumberToObject(body, DATA_FIELD_AQI, TUINT8);
+    JAddNumberToObject(body, DATA_FIELD_TVOC, TUINT16);
+    JAddNumberToObject(body, DATA_FIELD_ECO2, TUINT16);
+    JAddNumberToObject(body, DATA_FIELD_ETOH, TUINT16);
+    JAddNumberToObject(body, DATA_FIELD_TEMP, TFLOAT16);
+    JAddNumberToObject(body, DATA_FIELD_HUMIDITY, TFLOAT16);
+
+    // Attach the body to the request, and send it to the gateway
+    JAddItemToObject(req, "body", body);
+    noteSendToGatewayAsync(req, true);
+    return true;
+}
+
+// Gateway Response handler
+void aqiResponse(int appID, J *rsp, void *appContext)
+{
+
+    // Unused parameter(s)
+    (void)appID;
+    (void)appContext;
+
+    // See if there's an error
+    char *err = JGetString(rsp, "err");
+    if (err[0] != '\0') {
+        APP_PRINTF("aqi: gateway returned error: %s\r\n", err);
+        return;
+    }
+
+    switch (JGetInt(rsp, "id")) {
+
+    case REQUESTID_TEMPLATE:
```

Comment:
An if seems simpler here than a switch with only one case.

Reply:
I do agree - this was copied verbatim from nf33.